### DISCUSSION TITLE
Allow narrowing return types of interface members, like with overriden members

### DISCUSSION THREAD

**Discussion by RenderMichael** [dotnet/csharplang#7517] [Language Ideas]
Today in C#, an abstract class's member can be overridden with a narrower return type.

However, the same (or similar) is not allowed when implementing an interface.

Can this apparent inconsistency be made consistent, or is this by design/insurmountable?

```csharp
interface IMyInterface
{
    object Data2 { get; }
}

abstract class MyClass
{
    public abstract object Data1 { get; }
}

class Impl : MyClass, IMyInterface
{
    public override string Data1 => ""; // Legal

    public string Data2 => ""; // Illegal
}
```

**Comment by RikkiGibson** [comment]
Having not thought too deeply about it, it seems like the main challenge would be in getting the runtime to recognize when 'Impl.Data2' is implicitly implementing 'IMyInterface.Data2'. I am not sure if any compat problems could occur with that. Presumably it just changes the answer of "does this member implement the interface member" from "no" to "yes" in certain cases.

It's possible this capability wasn't considered to be as urgent to provide because the user can already declare an explicit implementation like
`object IMyInterface.Data2 => Data2;` in `Impl`.

**Reply by 333fred** [reply]
IIRC there were some big concerns about breaking changes and accidental overrides occurring when we looked at this. Unlike overriding class members, which must be virtual, that does not need to be the case for interface implementations, and there was the concern that upgrading could cause unexpected implementations and breaking changes in user code that may not even be visible until some unexpected runtime behavior occurs.

**Reply by RikkiGibson** [reply]
Yeah it sounds on the order of an overload resolution breaking change. If a more distant base has the exact return type, and a nearer base has a covariant return, we would be changing which method is chosen.

**Reply by RenderMichael** [reply]
Sorry i’m having trouble visualizing this. How can a method overload on return type?

**Reply by HaloFour** [reply]
The proposal states a case where the new runtime behavior would be a breaking change:

https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/covariant-returns#implicit-interface-implementations

```cs
using System;

interface I1 { object M(); }
class C1 : I1 { public object M() { return "C1.M"; } }
class C2 : C1, I1 { public new string M() { return "C2.M"; } }
class Program
{
    static void Main()
    {
        I1 i = new C2();
        Console.WriteLine(i.M());
    }
}
```

This is legal and prints "C1.M" today, but with the change it would print "C2.M" instead.

Note that actually wiring up implicit implementations is a part of the runtime, not the C# language.

**Reply by TahirAhmadov** [reply]
@333fred given this legacy, can we introduce new syntax to opt-in to this?
```cs
interface IMyInterface
{
    object Data2 { get; }
    object Foo(int x);
    object Foo(DateTime y);
}

abstract class MyClass
{
    public abstract object Data1 { get; }
}

class Impl : MyClass, IMyInterface
{
    public override string Data1 => ""; // Legal

    public string Data2 : IMyInterface.Data2 => ""; // legal now!
    public string Foo(int x) : IMyInterface.Foo => ""; // legal now! we use the parameter list to determine the overload
    public string Foo(DateTime y) : IMyInterface.Foo => ""; // legal now! we use the parameter list to determine the overload
// or
    public string IMyInterface.Data2 => ""; // legal now!
    public string IMyInterface.Foo(int x) => ""; // legal now! we use the parameter list to determine the overload
    public string IMyInterface.Foo(DateTime y) => ""; // legal now! we use the parameter list to determine the overload
}
```

**Reply by HaloFour** [reply]
See: #7210

I'm not sure that helps, tho, unless the runtime explicitly doesn't support covariant returns for implicit implementation and requires the `.overrides` modifier on the method in IL.

**Reply by RenderMichael** [reply]
I like the syntax mentioned in #7210 where you just add an accessibility modifier to an explicit implementation.

Maybe as a more mild approach, C# can forsake `: NewName` and require keeping the same name? That would allow narrowing return types (which could avoid boxing) without allowing a different method name (which doesn't prevent anything besides an additional method call, a la `Close() => Dispose()`, which can be optimized by the JIT)?

**Reply by HaloFour** [reply]
Note that with explicit implementations the interface is a part of the method name, e.g. `System.IDisposable.Dispose`.

```
    .method private final hidebysig newslot virtual 
        instance void System.IDisposable.Dispose () cil managed 
    {
        .override method instance void [System.Runtime]System.IDisposable::Dispose()

        // ...
    }
```
